### PR TITLE
Lesson 4: Homework - Van Truong

### PR DESCRIPTION
**Pull request by tadashii1417**

### CODE REVIEW

File: homework/Truong_Duong/lesson_4/javascript_exercise.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+```
+
+A: Lydia and undefined
+B: Lydia and ReferenceError
+C: ReferenceError and 21
+D: undefined and ReferenceError
+
+> **Answer: D**
+> Explain: variable defined with `var` was hoisted the `declaration part` into the begining of the function, but the value did not. Therefore, it returned `undefined` with `name`
+> While with variable defined with `let`, a ES6 syntax, it wasn't hoisted by default. Therefore, it returned a ReferenceError here.
```
Comment (trongtai37) [inline homework/Truong_Duong/lesson_4/javascript_exercise.md:23]: I think when using `let` or `const` to declare variables, it still be hoisted but we should notice about the TDZ (temporal dead zone)
You can read more here: https://developer.mozilla.org/en-US/docs/Glossary/Hoisting?retiredLocale=vi#let_and_const_hoisting

Comment (tadashii1417) [reply homework/Truong_Duong/lesson_4/javascript_exercise.md:23]: @trongtai37 Thank you for your information 🙇

File: homework/Truong_Duong/lesson_4/javascript_exercise.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+A: true false true
+B: false false true
+C: true false false
+D: false true true
+
+> **Answer: C**
+> Reference here: [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)
+> In short, `Number('123') === 123`
+> When Number is called as a constructor (with new), it creates a `Number object`, which is not a primitive. `typeof new Number(42) === "object"`, and `new Number(42) !== 42`
```
Comment (trongtai37) [inline homework/Truong_Duong/lesson_4/javascript_exercise.md:140]: Honestly, I have never use `new Number()` or `new Boolean(), new String(), ...`. I get used to using these classes for casting variables like `Number('3') or Boolean(-1)` :D
This information help me a lot. Thanks

Comment (trongtai37) [inline homework/Truong_Duong/lesson_4/javascript_exercise.md:140]: Honestly, I have never use `new Number()` or `new Boolean(), new String(), ...`. I get used to using these classes for casting variables like `Number('3') or Boolean(-1)` :D
This information help me a lot. Thanks

File: homework/Truong_Duong/lesson_4/53_Maximum_subarray.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * @return {number}
+ */
+ var maxSubArray = function(nums) {
+    const size = nums.length;
+    const prefixSum = new Array(size).fill(0);
+    
+    let ans = Number.MIN_SAFE_INTEGER, min = 0;
+          
+    for (let i = 0; i < size; i++) {
+        prefixSum[i] = i == 0 ? nums[i] : prefixSum[i-1] + nums[i];
```
Comment (quanghoang33) [inline homework/Truong_Duong/lesson_4/53_Maximum_subarray.js:14]: recommend: in a coding interview, I would not recommend using ternary operation, it is hard to read/ debug

**Review by quanghoang33 — APPROVED** [review]

**Pull request merged**